### PR TITLE
Fixed #24877 -- Call process_exception on exceptions raised in render

### PR DESCRIPTION
**Pull request by sephii**

### CODE REVIEW

File: django/core/handlers/base.py [hunk 1]
```diff
@@ -105,6 +105,15 @@ def get_exception_response(self, request, resolver, status_code, exception):
 
     def get_response(self, request):
         "Returns an HttpResponse object for the given HttpRequest"
+        def process_exception_by_middlewares(exception):
```
Comment (timgraham) [inline django/core/handlers/base.py]: Is there any benefit to making this an inner method as opposed to putting it on the class?

File: tests/template_tests/test_response.py [hunk 2]
```diff
@@ -221,6 +221,12 @@ def test_pickling_cookie(self):
 
         self.assertEqual(unpickled_response.cookies['key'].value, 'value')
 
+    @override_settings(ROOT_URLCONF='template_tests.urls',
```
Comment (timgraham) [inline tests/template_tests/test_response.py]: `template_tests` isn't the correct place for this test (it has nothing to do with the template system). Please look around for  a better place (if you need help or want to confirm before moving things, let me know).

**Comment by timgraham** [comment]
This change should be mentioned in the 1.9 release notes (not sure about elsewhere in the documentation).

File: tests/template_tests/test_response.py (hunk 2 above)
Comment (sephii) [inline tests/template_tests/test_response.py]: Haha, I was thinking about putting it in the handlers tests and @freakboy3742 told me to put that here. ;)

File: django/core/handlers/base.py (hunk 1 above)
Comment (sephii) [inline django/core/handlers/base.py]: I just thought this was specific enough to go in an inner method. I just changed it and put it on the class.

File: tests/template_tests/test_response.py (hunk 2 above)
Comment (timgraham) [inline tests/template_tests/test_response.py]: How about `tests/middleware_exceptions`?

Comment (sephii) [inline tests/template_tests/test_response.py]: Ok moved to `tests/middleware_exceptions`.

**Comment by sephii** [comment]
I added a change in the release notes. Please let me know if it's in the right place in the file.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/middleware_exceptions/views.py [hunk 3]
```diff
@@ -32,3 +32,12 @@ def null_view(request):
 
 def permission_denied(request):
     raise PermissionDenied()
+
+
+def exception_in_render(request):
+    class CustomHttpResponse(http.HttpResponse):
+        def render(self):
+            raise Exception('Exception in HttpResponse render that should be '
```
Comment (timgraham) [inline tests/middleware_exceptions/views.py]: preferred style, also "catched" -> "caught"
```
raise Exception(
    'Exception... that should be caught'
    '...' 
)
```

File: tests/middleware_exceptions/tests.py [hunk 4]
```diff
@@ -486,6 +486,15 @@ def test_process_template_response_error(self):
         # Check that the right middleware methods have been invoked
         self.assert_middleware_usage(middleware, True, True, True, True, False)
 
+    @override_settings(ROOT_URLCONF='middleware_exceptions.urls',
```
Comment (timgraham) [inline tests/middleware_exceptions/tests.py]: preferred style:
```
@override_settings(
   ROOT_URLCONF=...
)
```
although ROOT_URLCONF is already specified on the class level -- does it need to be repeated?

File: tests/middleware_exceptions/middleware.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+from django.http import HttpResponse
```
Comment (timgraham) [inline tests/middleware_exceptions/middleware.py]: I think adding `from __future__ import unicode_literals` to this file will fix the test failure on Python 3.

File: tests/middleware_exceptions/middleware.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+
+
+class ProcessExceptionMiddleware(object):
+    def process_exception(self, request, exception):
+        return HttpResponse('Exception catched')
```
Comment (timgraham) [inline tests/middleware_exceptions/middleware.py]: catched -> caught

File: docs/releases/1.9.txt [hunk 7]
```diff
@@ -319,6 +319,9 @@ Requests and Responses
 * The default 40x error views now accept a second positional parameter, the
   exception that triggered the view.
 
+* Exceptions raised by the :meth:`render() <django.template.response>` method
+  are now correctly passed to the middlewares ``process_exception`` methods.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: "the render method of template responses" (chop link - not sure it adds clarifty?)
chop "correctly"?
passed to the process_exception() method of each middleware.

**Comment by sephii** [comment]
I'm not sure how to fix the following failing test: http://djangoci.com/job/pull-requests-trusty/2248/database=sqlite3,label=trusty-pr,python=python3.4/testReport/middleware_exceptions.tests/MiddlewareTests/test_exception_in_render_passed_to_process_exception/

Adding `from __future__ import unicode_literals` doesn't seem to have fixed the problem. I guess that's because the text of the exception is a string, while `response.content` is always binary? Should I `force_text(response.content)` in the assertion?

**Comment by timgraham** [comment]
You can add a b'' prefix: `self.assertEqual(response.content, b'Exception caught')`

**Comment by sephii** [comment]
I added the `b` prefix as you can see in the PR code, but the tests keep failing for some reason: http://djangoci.com/job/pull-requests-trusty/2257/

As you can see on the test fail page it looks like it's still referencing the previous version of the code (`self.assertEqual(response.content, 'Exception caught')`). I ran the tests locally on Python 3 and they're running fine.

**Comment by timgraham** [comment]
buildbot, test this please.

**Comment by timgraham** [comment]
buildbot, test this please.

**Comment by timgraham** [comment]
merged in f5d5867a4a6aeddd58ff855a01ab4e438d938ac1, thanks!

**Pull request closed without merging**